File: Data Structures & Algorithms/mergeSort/submission-1.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
    vector<Pair> merge(vector<Pair>& left, vector<Pair>& right) {
        // PUSH THE SMALLEST ELEMENT FROM CURRENT PTR TO SORTED
        vector<Pair> sorted;
        int lPtr = 0;
        int rPtr = 0;
        while (lPtr < left.size() && rPtr < right.size()) {
            if (left[lPtr].key <= right[rPtr].key) {
                sorted.push_back(left[lPtr]);
                lPtr++;
            }
            else {
                sorted.push_back(right[rPtr]);
                rPtr++;
            }
        }
        // GET REST OF UNFINISHED ARRAY
        while (rPtr < right.size()) {
            sorted.push_back(right[rPtr]);
            rPtr++;
        }
        while (lPtr < left.size()) {
            sorted.push_back(left[lPtr]);
            lPtr++;
        }
        return sorted;
    }

    vector<Pair> mergeSort(vector<Pair>& pairs) {
        // BASE CASE
        if (pairs.size() < 2) {
            return pairs;
        }
        // SORT LEFT & RIGHT HALF, THEN MERGE THEM
        int mid = pairs.size() / 2;
        vector<Pair> left(pairs.begin(), pairs.begin() + mid);
        vector<Pair> right(pairs.begin() + mid, pairs.end());
        vector<Pair> leftSort = mergeSort(left);
        vector<Pair> rightSort = mergeSort(right);
        return merge(leftSort, rightSort);
    }
};

File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        lptr = 0
        rptr = len(s) - 1 

        while lptr < rptr and rptr > 0 and lptr < len(s):
            while not s[lptr].isalnum() and lptr < rptr:
                lptr += 1
            while not s[rptr].isalnum() and rptr > lptr:
                rptr -= 1
            
            if s[lptr].lower() != s[rptr].lower() and lptr < rptr:
                return False
            
            lptr += 1
            rptr -= 1
        
        return True